### PR TITLE
Nested link fix

### PR DESCRIPTION
**Pull request by rynomad**
links inside `<strong>` weren't working, now they are.

e.g

```
<strong><a href="http://example.com">link me</a></strong> 
- was doing -> **link me**
- now doing -> **[link me](http://example.com)**
```

added a test case too.

### CODE REVIEW

**Comment by rynomad** [comment]
oops, accidentally cluttered git history with bundle, I reverted it but can do a cleaner PR if it's a big deal

File: src/core/markdownASTToString.ts [hunk 1]
```diff
@@ -82,35 +82,48 @@ function markdownContentASTToString(nodes: SemanticMarkdownAST[], options?: Conv
...
+                        }
                         markdownString += `${indent}${node.content}`;
-                    } else if (node.type === 'bold') {
-                        markdownString += `**${node.content}**`;
-                    } else if (node.type === 'italic') {
-                        markdownString += `*${node.content}*`;
-                    } else if (node.type === 'strikethrough') {
-                        markdownString += `~~${node.content}~~`;
-                    } else if (node.type === 'link') {
-                        // Check if the link contains only text
```
Comment (romansky) [inline src/core/markdownASTToString.ts:103]: please add these comments back

File: src/core/markdownASTToString.ts [hunk 2]
```diff
@@ -82,35 +82,48 @@ function markdownContentASTToString(nodes: SemanticMarkdownAST[], options?: Conv
...
-                        if (node.content.length === 1 && node.content[0].type === 'text') {
-                            // Use native markdown syntax for text-only links
```
Comment (romansky) [inline src/core/markdownASTToString.ts:105]: please add these comments back

File: src/core/markdownASTToString.ts [hunk 3]
```diff
@@ -82,35 +82,48 @@ function markdownContentASTToString(nodes: SemanticMarkdownAST[], options?: Conv
...
-                            markdownString += ` [${node.content[0].content}](${node.href})`;
+                    } else {
+                        let content = '';
+                        if (Array.isArray(node.content)) {
+                            content = markdownContentASTToString(node.content, options, indentLevel);
                         } else {
-                            // Use HTML <a> tag for links with rich content
```
Comment (romansky) [inline src/core/markdownASTToString.ts:108]: please add these comments back

File: src/core/markdownASTToString.ts [hunk 4]
```diff
@@ -82,35 +82,48 @@ function markdownContentASTToString(nodes: SemanticMarkdownAST[], options?: Conv
...
-                            const linkContent = markdownContentASTToString(node.content, options, indentLevel + 1);
-                            markdownString += ` <a href="${node.href}">${linkContent}</a>`;
+                            content = node.content;
                         }
-                    }
+                        content = content.trim();
 
+                        if (!isPrevNodeFormatting && markdownString.length > 0 && !/\s$/.test(markdownString)) {
```
Comment (romansky) [inline src/core/markdownASTToString.ts:104]: /\s$/.test(markdownString) is an O(n) operation and could be quite expensive with large documents.. please use /\s/.test(markdownString.slice(-1)) instead.
Also, please use the value from the dup code above..

File: src/core/markdownASTToString.ts [hunk 5]
```diff
@@ -82,35 +82,48 @@ function markdownContentASTToString(nodes: SemanticMarkdownAST[], options?: Conv
...
-                        markdownString += ' ';
-                    }
+                    const prevNode = index > 0 ? nodes[index - 1] : null;
+                    const nextNode = index < nodes.length - 1 ? nodes[index + 1] : null;
+                    const isPrevNodeFormatting = prevNode && ['bold', 'italic', 'strikethrough', 'link'].includes(prevNode.type);
+                    const isNextNodePunctuation = nextNode && nextNode.type === 'text' && /^[.,!?;:]/.test(nextNode.content);
 
                     if (node.type === 'text') {
-
+                        if (!isPrevNodeFormatting && !isNextNodePunctuation && !/^\s/.test(node.content) && markdownString.length > 0 && !/\s$/.test(markdownString)) {
```
Comment (romansky) [inline src/core/markdownASTToString.ts:91]: /\s$/.test(markdownString) is an O(n) operation and could be quite expensive with large documents.. please use /\s/.test(markdownString.slice(-1) instead.
Also, please move it a bit up and give this a meaningful name

File: src/core/markdownASTToString.ts [hunk 6]
```diff
@@ -82,35 +82,48 @@ function markdownContentASTToString(nodes: SemanticMarkdownAST[], options?: Conv
...
+                        } else if (node.type === 'link') {
+                            if (node.content.length === 1 && node.content[0].type === 'text') {
+                                markdownString += `[${content}](${node.href})`;
+                            } else {
+                                markdownString += `<a href="${node.href}">${content}</a>`;
+                            }
+                        }
+
+                        // Add space after closing tag if next node is not punctuation
+                        if (!isNextNodePunctuation && nextNode) {
```
Comment (romansky) [inline src/core/markdownASTToString.ts:123]: Not sure this is needed.. the next node should add it's own space in the beginning if it's needed

**Review by romansky — CHANGES_REQUESTED** [review]
Great pull request! please fix the few issues and I'm happy to merge.

**Comment by rynomad** [comment]
roger, will clean it up this evening!

File: src/core/markdownASTToString.ts (hunk 6 above)
Comment (romansky) [reply src/core/markdownASTToString.ts:129]: Please explain why this is needed..

**Review by romansky — CHANGES_REQUESTED** [review]
Almost there :)

**Comment by rynomad** [comment]
Yeah sorry about the delay, was trying to get it over the line while waiting for a batch job to finish but got sucked back into it before i could finish, and my brain is currently mush for at least another day or so. Will complete asap. Love this lib btw, got a ton of other ideas to pitch you on. Expect more PRs :)

**Comment by rynomad** [comment]
update: haven't forgotten, just swamped... might not be able to sit down with this until the weekend. In the spirit of honesty, claude wrote the initial PR because I just needed the problem solved and I'm using esm.sh pointed at my forked repo to unblock my work. I suppose this makes a good little case study on llm codegen being "technically correct" without regard for the bigger picture.

**Comment by rynomad** [comment]
closing in favor of [new pr](https://github.com/romansky/dom-to-semantic-markdown/pull/17)

**Pull request closed without merging**